[BUG] Resource identifier doesn't check existing indentifiers

### Is there an existing Issue for this bug?

- [X] I have searched the existing issues

### Expected behavior

When generating a unique key for an image a check should be performed to see if any previous images have generated the same key.

### Actual behavior

A duplicate key can be generated resulting in the image upload failing.

### Steps to reproduce

1. Screenshot and upload to ass.
2. Repeat step 1 until logs show `Error: Key already exists`.

### Logs (optional)

```text
Oct 2, 2021, 12:39 AM [ERROR]    Error: Key already exists
Error: Key already exists
    at @tycrek/papito/JsonDataEngine.js:66:13
    at new Promise (<anonymous>)
    at JsonPutFunc (@tycrek/papito/JsonDataEngine.js:63:9)
    at JsonDataEngine.put (@tycrek/papito/DataEngine.js:55:29)
    at /routers/upload.js:70:7
    at Layer.handle [as handle_request] (express/lib/router/layer.js:95:5)
    at next (express/lib/router/route.js:137:13)
    at Route.dispatch (express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (express/lib/router/layer.js:95:5)
    at express/lib/router/index.js:281:22
Oct 2, 2021, 12:39 AM [Express] Response: 500
```

### ass version

0.8.6

### Deployment

Manual

### File storage

Local

### Operating system

Ubuntu 20.04.2 LTS

### Node version

14.17.1

### Reverse proxy

- [X] I am using a reverse proxy

### Reverse proxy software

nginx/1.18.0 (Ubuntu) 

### Cloudflare

- [ ] I am using Cloudflare

 👍 thanks for reporting, I'll look into that soon

Oct. 1, 2021, 8:52 a.m. by ***@***.***:

>
>
>
> Is there an existing Issue for this bug?
> I have searched the existing issues
> Expected behavior
>
> When generating a unique key for an image a check should be performed to see if any previous images have generated the same key.
>
> Actual behavior
>
> A duplicate key can be generated resulting in the image upload failing.
>
> Steps to reproduce
> Screenshot
> Repeat step 1 until logs show > Error: Key already exists
> Logs (optional)
> Oct 2, 2021, 12:39 AM [ERROR]    Error: Key already existsError: Key already exists    at @tycrek/papito/JsonDataEngine.js:66:13    at new Promise (<anonymous>)    at JsonPutFunc ***@***.***/papito/JsonDataEngine.js:63:9)    at JsonDataEngine.put ***@***.***/papito/DataEngine.js:55:29)    at /routers/upload.js:70:7    at Layer.handle [as handle_request] (express/lib/router/layer.js:95:5)    at next (express/lib/router/route.js:137:13)    at Route.dispatch (express/lib/router/route.js:112:3)    at Layer.handle [as handle_request] (express/lib/router/layer.js:95:5)    at express/lib/router/index.js:281:22Oct 2, 2021, 12:39 AM [Express] Response: 500
> ### ass version0.8.6### DeploymentManual### File storageLocal### Operating systemUbuntu 20.04.2 LTS### Node version14.17.1### StorageEngineLocal storage### Reverse proxy- [X] I am using a reverse proxy### Reverse proxy softwarenginx/1.18.0 (Ubuntu) ### Cloudflare- [ ] I am using Cloudflare
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, > view it on GitHub <https://github.com/tycrek/ass/issues/75>> , or > unsubscribe <https://github.com/notifications/unsubscribe-auth/AHEKGAFGNUYNPBC7ES4ABGDUEXDJNANCNFSM5FE32UPA>> .
>
>
>

--- a/routers/upload.js
+++ b/routers/upload.js
@@ -64,62 +64,80 @@ router.post('/', (req, res, next) => {
 	// Fix spaces in originalname
 	req.file.originalname = req.file.originalname.replace(/\s/g, spaceReplace === '!' ? '' : spaceReplace);
 
-	// Save the file information
-	const resourceId = generateId(generator, resourceIdSize, req.headers['x-ass-gfycat'] || gfyIdSize, req.file.originalname);
-	log.debug('Saving data', data.name);
-	data.put(resourceId.split('.')[0], req.file).then(() => {
-		// Log the upload
-		const logInfo = `${req.file.originalname} (${req.file.mimetype}, ${formatBytes(req.file.size)})`;
-		log.success('File uploaded', logInfo, `uploaded by ${users[req.token] ? users[req.token].username : '<token-only>'}`);
+	// Generate a unique resource ID
+	let resourceId;
 
-		// Build the URLs
-		const resourceUrl = `${getTrueHttp()}${trueDomain}/${resourceId}`;
-		const thumbnailUrl = `${getTrueHttp()}${trueDomain}/${resourceId}/thumbnail`;
-		const deleteUrl = `${getTrueHttp()}${trueDomain}/${resourceId}/delete/${req.file.deleteId}`;
+	// Function to call to generate a fresh ID. Used for multiple attempts in case an ID is already taken
+	const gen = () => generateId(generator, resourceIdSize, req.headers['x-ass-gfycat'] || gfyIdSize, req.file.originalname);
 
-		// Send the response
-		res.type('json').send({ resource: resourceUrl, thumbnail: thumbnailUrl, delete: deleteUrl })
-			.on('finish', () => {
-				log.debug('Upload response sent');
+	// Called by a promise, this will recursively resolve itself until a unique ID is found
+	// TODO: Add max attempts in case all available ID's are taken
+	function genCheckId(resolve, reject) {
+		let uniqueId = gen();
+		data.has(uniqueId)
+			.then((exists) => (log.debug('ID check', exists ? 'Taken' : 'Available'), exists ? genCheckId(resolve, reject) : resolve(uniqueId)))
+			.catch(reject);
+	}
 
-				// After we have sent the user the response, also send a Webhook to Discord (if headers are present)
-				if (req.headers['x-ass-webhook-client'] && req.headers['x-ass-webhook-token']) {
+	new Promise((resolve, reject) => genCheckId(resolve, reject))
+		.then((uniqueId) => {
+			resourceId = uniqueId;
+			log.debug('Saving data', data.name);
+		})
+		.then(() => data.put(resourceId.split('.')[0], req.file))
+		.then(() => {
+			// Log the upload
+			const logInfo = `${req.file.originalname} (${req.file.mimetype}, ${formatBytes(req.file.size)})`;
+			log.success('File uploaded', logInfo, `uploaded by ${users[req.token] ? users[req.token].username : '<token-only>'}`);
 
-					// Build the webhook
-					const hook = new Webhook(req.headers['x-ass-webhook-url']);
-					hook.setUsername(req.headers['x-ass-webhook-username'] || 'ass');
-					hook.setAvatar(req.headers['x-ass-webhook-avatar'] || ASS_LOGO);
+			// Build the URLs
+			const resourceUrl = `${getTrueHttp()}${trueDomain}/${resourceId}`;
+			const thumbnailUrl = `${getTrueHttp()}${trueDomain}/${resourceId}/thumbnail`;
+			const deleteUrl = `${getTrueHttp()}${trueDomain}/${resourceId}/delete/${req.file.deleteId}`;
 
-					// Build the embed
-					const embed = new MessageBuilder()
-						.setTitle(logInfo)
-						.setURL(resourceUrl)
-						.setDescription(`**Size:** \`${formatBytes(req.file.size)}\`\n**[Delete](${deleteUrl})**`)
-						.setThumbnail(thumbnailUrl)
-						.setColor(req.file.vibrant)
-						.setTimestamp();
+			// Send the response
+			res.type('json').send({ resource: resourceUrl, thumbnail: thumbnailUrl, delete: deleteUrl })
+				.on('finish', () => {
+					log.debug('Upload response sent');
 
-					// Send the embed to the webhook, then delete the client after to free resources
-					log.debug('Sending embed to webhook');
-					hook.send(embed)
-						.then(() => log.debug('Webhook sent'))
-						.catch((err) => log.error('Webhook error').err(err));
-				}
+					// After we have sent the user the response, also send a Webhook to Discord (if headers are present)
+					if (req.headers['x-ass-webhook-client'] && req.headers['x-ass-webhook-token']) {
 
-				// Also update the users upload count
-				if (!users[req.token]) {
-					const generateUsername = () => generateId('random', 20, null); // skipcq: JS-0074
-					let username = generateUsername();
-					while (Object.values(users).findIndex((user) => user.username === username) !== -1)  // skipcq: JS-0073
-						username = generateUsername();
-					users[req.token] = { username, count: 0 };
-				}
-				users[req.token].count += 1;
-				fs.writeJsonSync(path('auth.json'), { users }, { spaces: 4 });
+						// Build the webhook
+						const hook = new Webhook(req.headers['x-ass-webhook-url']);
+						hook.setUsername(req.headers['x-ass-webhook-username'] || 'ass');
+						hook.setAvatar(req.headers['x-ass-webhook-avatar'] || ASS_LOGO);
 
-				log.debug('Upload request flow completed', '');
-			});
-	}).catch(next);
+						// Build the embed
+						const embed = new MessageBuilder()
+							.setTitle(logInfo)
+							.setURL(resourceUrl)
+							.setDescription(`**Size:** \`${formatBytes(req.file.size)}\`\n**[Delete](${deleteUrl})**`)
+							.setThumbnail(thumbnailUrl)
+							.setColor(req.file.vibrant)
+							.setTimestamp();
+
+						// Send the embed to the webhook, then delete the client after to free resources
+						log.debug('Sending embed to webhook');
+						hook.send(embed)
+							.then(() => log.debug('Webhook sent'))
+							.catch((err) => log.error('Webhook error').err(err));
+					}
+
+					// Also update the users upload count
+					if (!users[req.token]) {
+						const generateUsername = () => generateId('random', 20, null); // skipcq: JS-0074
+						let username = generateUsername();
+						while (Object.values(users).findIndex((user) => user.username === username) !== -1)  // skipcq: JS-0073
+							username = generateUsername();
+						users[req.token] = { username, count: 0 };
+					}
+					users[req.token].count += 1;
+					fs.writeJsonSync(path('auth.json'), { users }, { spaces: 4 });
+
+					log.debug('Upload request flow completed', '');
+				});
+		}).catch(next);
 });
 
 module.exports = router;

--- a/routers/upload.js
+++ b/routers/upload.js
@@ -65,7 +65,7 @@ router.post('/', (req, res, next) => {
 	req.file.originalname = req.file.originalname.replace(/\s/g, spaceReplace === '!' ? '' : spaceReplace);
 
 	// Generate a unique resource ID
-	let resourceId;
+	let resourceId = '';
 
 	// Function to call to generate a fresh ID. Used for multiple attempts in case an ID is already taken
 	const gen = () => generateId(generator, resourceIdSize, req.headers['x-ass-gfycat'] || gfyIdSize, req.file.originalname);
@@ -73,9 +73,9 @@ router.post('/', (req, res, next) => {
 	// Called by a promise, this will recursively resolve itself until a unique ID is found
 	// TODO: Add max attempts in case all available ID's are taken
 	function genCheckId(resolve, reject) {
-		let uniqueId = gen();
+		const uniqueId = gen();
 		data.has(uniqueId)
-			.then((exists) => (log.debug('ID check', exists ? 'Taken' : 'Available'), exists ? genCheckId(resolve, reject) : resolve(uniqueId)))
+			.then((exists) => (log.debug('ID check', exists ? 'Taken' : 'Available'), exists ? genCheckId(resolve, reject) : resolve(uniqueId))) // skipcq: JS-0090
 			.catch(reject);
 	}
 

--- a/routers/upload.js
+++ b/routers/upload.js
@@ -70,12 +70,21 @@ router.post('/', (req, res, next) => {
 	// Function to call to generate a fresh ID. Used for multiple attempts in case an ID is already taken
 	const gen = () => generateId(generator, resourceIdSize, req.headers['x-ass-gfycat'] || gfyIdSize, req.file.originalname);
 
+	// Keeps track of the number of attempts in case all ID's are taken
+	const attempts = {
+		count: 0,
+		max: 50
+	};
+
 	// Called by a promise, this will recursively resolve itself until a unique ID is found
-	// TODO: Add max attempts in case all available ID's are taken
 	function genCheckId(resolve, reject) {
 		const uniqueId = gen();
+		attempts.count++;
 		data.has(uniqueId)
-			.then((exists) => (log.debug('ID check', exists ? 'Taken' : 'Available'), exists ? genCheckId(resolve, reject) : resolve(uniqueId))) // skipcq: JS-0090
+			.then((exists) => {
+				log.debug('ID check', exists ? 'Taken' : 'Available');
+				return attempts.count - 1 >= attempts.max ? reject(new Error('No ID\'s remaining')) : exists ? genCheckId(resolve, reject) : resolve(uniqueId);
+			})
 			.catch(reject);
 	}
 

